### ISSUE TITLE
Implement/Port Outbound ACL policy module

### ISSUE THREAD

**Issue by efimki** [envoyproxy/envoy#10336]
*Title*: *Implement/Port Outbound ACL policy module*

*Description*:
We need to be able to set ranges of IP addresses and ports that Envoy is allowed to connect to. 
For example, proxy should be able to send requests to 1.2.3.4:80 but not to 1.2.3.4:22 or 5.6.7.8:80.

Per conversation with alyssar@ this should probably be intercepted at router filter, but is there some short-term workaround that we can use?

**Comment by mattklein123** [comment]
@efimki in general Envoy does not allow the downstream client to pick who they connect to, it is configured by the management server/config. There are some exceptions to this in a few places, but can you clarify which case you care about specifically? That might help scope what we need to do here. Thank you!

**Comment by efimki** [comment]
@mattklein123 Unfortunately I'm not totally fluent in Envoy terminology, but hopefully @alyssawilk could help me to explain.

Here is the use case I can think of:
1. Management server/config is configured to connect to "https://www.badexample.com"
2. The owner of "badexample.com" changes DNS record to resolve to the address local to the network that runs Envoy (let's say 5.6.7.8) the owner of 'badexample.com' shouldn't have access to.
3. The client sends request to Envoy gets routed to 5.6.7.8. This is problematic.

I would like to be able to specify that Envoy is allowed or denied connection to particular IP range. 
In this example I could mark the range 5.0.0.0/8 as 'denied' and Envoy would NOT connect to it regardless of server/config.

Does this make sense? Maybe there is some way to achieve this already?

**Comment by mattklein123** [comment]
> Does this make sense? Maybe there is some way to achieve this already?

Yes it makes sense, but it's a complex issue that could have multiple different solutions. For example:
1) If DNS is the primary issue, could we actually put the blocking mechanism in the DNS resolution code as a list of banned CIDRs? This would cover dynamic HTTP proxy, static_dns, logical_dns, etc.
2) If we want a more general blocking solution, I could imagine a few different ways of doing this, but the "simplest" way is probably an upstream connection filter that closes the connection if it is attempted to a bad IP?

**Comment by efimki** [comment]
> Yes it makes sense, but it's a complex issue that could have multiple different solutions. 

Excellent! 

> 1. If DNS is the primary issue, could we actually put the blocking mechanism in the DNS resolution code as a list of banned CIDRs? This would cover dynamic HTTP proxy, static_dns, logical_dns, etc.

I think DNS is an important example, but limiting on the DNS level seems a bit odd as we would still need to do the same checks on connection level to cover the case when bad IP is configured explicitly.

> 2. If we want a more general blocking solution, I could imagine a few different ways of doing this, but the "simplest" way is probably an upstream connection filter that closes the connection if it is attempted to a bad IP?

The upstream connection filter that closes the connection sounds like a right direction, although it would be nice to prevent connection from being established in the first place. 
I can imagine some "interesting" probing / timing attacks that explore network topology by looking at time to receive an error...

**Comment by mattklein123** [comment]
> The upstream connection filter that closes the connection sounds like a right direction, although it would be nice to prevent connection from being established in the first place.
I can imagine some "interesting" probing / timing attacks that explore network topology by looking at time to receive an error...

Unfortunately I don't think there is going to be an easy answer here so I'm marking this as a design proposal needed. Some things to think about:

1. An upstream connection filter will definitely work, but as you say it will happen pretty late in the process. Note that I think with the filter we can prevent the connection from being established at all so I don't think there would be a probing issue. The pro of this approach is that it will apply to any kind of connection.
2. I don't think doing something in the HTTP router makes sense, since the logic would have to be replicated in the tcp_proxy, redis, etc. (if you use other protocols) so I think we need to do this in a more central place.
3. If an upstream connection filter doesn't seem right, the other thing that comes to mind is some kind of extensible cluster endpoint filtering system, where every time a cluster gets updated, you have the opportunity to prune the endpoints. I think this would be a pretty clean solution to the problem which would allow you to fully control the filtering logic.

cc @snowp for other thoughts on this.

**Comment by snowp** [comment]
What do we want the behavior of a rejected connection to be? Should we be attempting retries? I can imagine yes in certain cases, for example if the concern is corrupted/invalid service discovery data and not a bad actor.

I think having this happen in an upstream network filter makes sense: the router would see the connection reset and retry if configured to do so. If the current interface doesn't allow us to stop the connection before it happens (is `onNewConnection` called before or after we attempt to connect?) we could probably extend that interface to have both a pre-connection and post-connection hook.

**Comment by mattklein123** [comment]
> If the current interface doesn't allow us to stop the connection before it happens (is onNewConnection called before or after we attempt to connect?) we could probably extend that interface to have both a pre-connection and post-connection hook.

Yeah that's what I was thinking as well.

**Comment by efimki** [comment]
AFAICT onNewConnection() is only called if connection_.state() == Connection::State::Open so I would guess it is called after we succeed to connect:

https://github.com/envoyproxy/envoy/blob/2c03496a6b5e9e32bc10db40b3fb85e4610bdde1/source/common/network/filter_manager_impl.cc#L39
